File: library/binpatch.cpp
/*
https://github.com/peterix/dfhack
Copyright (c) 2011 Petr Mrázek <peterix@gmail.com>

A thread-safe logging console with a line editor for windows.

Based on linenoise win32 port,
copyright 2010, Jon Griffiths <jon_p_griffiths at yahoo dot com>.
All rights reserved.
Based on linenoise, copyright 2010, Salvatore Sanfilippo <antirez at gmail dot com>.
The original linenoise can be found at: http://github.com/antirez/linenoise

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

  * Redistributions of source code must retain the above copyright notice,
    this list of conditions and the following disclaimer.
  * Redistributions in binary form must reproduce the above copyright
    notice, this list of conditions and the following disclaimer in the
    documentation and/or other materials provided with the distribution.
  * Neither the name of Redis nor the names of its contributors may be used
    to endorse or promote products derived from this software without
    specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.
*/


#include <stdarg.h>
#include <errno.h>
#include <stdio.h>
#include <assert.h>
#include <iostream>
#include <fstream>
#include <istream>
#include <string>
#include <stdint.h>

#include <cstdio>
#include <cstdlib>
#include <sstream>
#include <vector>

#include <memory>

#include <md5wrapper.h>

using std::cout;
using std::cerr;
using std::endl;

typedef unsigned char patch_byte;

struct BinaryPatch {
    struct Byte {
        unsigned offset;
        patch_byte old_val, new_val;
    };
    enum State {
        Conflict = 0,
        Unapplied = 1,
        Applied = 2,
        Partial = 3
    };

    std::vector<Byte> entries;

    bool loadDIF(std::string name);
    State checkState(const patch_byte *ptr, size_t len);

    void apply(patch_byte *ptr, size_t len, bool newv);
};

inline bool is_hex(char c)
{
    return (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F');
}

bool BinaryPatch::loadDIF(std::string name)
{
    entries.clear();

    std::ifstream infile(name);
    if(infile.bad())
    {
        cerr << "Cannot open file: " << name << endl;
        return false;
    }

    std::string s;
    while(std::getline(infile, s))
    {
        // Parse lines that begin with "[0-9a-f]+:"
        size_t idx = s.find(':');
        if (idx == std::string::npos || idx == 0 || idx > 8)
            continue;

        bool ok = true;
        for (size_t i = 0; i < idx; i++)
            if (!is_hex(s[i]))
                ok = false;
        if (!ok)
            continue;

        unsigned off, oval, nval;
        int nchar = 0;
        int cnt = sscanf(s.c_str(), "%x: %x %x%n", &off, &oval, &nval, &nchar);

        if (cnt < 3)
        {
            cerr << "Could not parse: " << s << endl;
            return false;
        }

        for (size_t i = nchar; i < s.size(); i++)
        {
            if (!isspace(s[i]))
            {
                cerr << "Garbage at end of line: " << s << endl;
                return false;
            }
        }

        if (oval >= 256 || nval >= 256)
        {
            cerr << "Invalid byte values: " << s << endl;
            return false;
        }

        Byte bv = { off, patch_byte(oval), patch_byte(nval) };
        entries.push_back(bv);
    }

    if (entries.empty())
    {
        cerr << "No lines recognized." << endl;
        return false;
    }

    return true;
}

BinaryPatch::State BinaryPatch::checkState(const patch_byte *ptr, size_t len)
{
    int state = 0;

    for (size_t i = 0; i < entries.size(); i++)
    {
        Byte &bv = entries[i];

        if (bv.offset >= len)
        {
            cerr << "Offset out of range: 0x" << std::hex << bv.offset << std::dec << endl;
            return Conflict;
        }

        patch_byte cv = ptr[bv.offset];
        if (bv.old_val == cv)
            state |= Unapplied;
        else if (bv.new_val == cv)
            state |= Applied;
        else
        {
            cerr << std::hex << bv.offset << ": "
                 << unsigned(bv.old_val) << " " << unsigned(bv.new_val)
                 << ", but currently " << unsigned(cv) << std::dec << endl;
            return Conflict;
        }
    }

    return State(state);
}

void BinaryPatch::apply(patch_byte *ptr, size_t len, bool newv)
{
    for (size_t i = 0; i < entries.size(); i++)
    {
        Byte &bv = entries[i];
        assert (bv.offset < len);

        ptr[bv.offset] = (newv ? bv.new_val : bv.old_val);
    }
}

bool load_file(std::vector<patch_byte> *pvec, std::string fname)
{
    FILE *f = fopen(fname.c_str(), "rb");
    if (!f)
    {
        cerr << "Cannot open file: " << fname << endl;
        return false;
    }

    fseek(f, 0, SEEK_END);
    pvec->resize(ftell(f));
    fseek(f, 0, SEEK_SET);
    size_t cnt = fread(pvec->data(), 1, pvec->size(), f);
    fclose(f);

    return cnt == pvec->size();
}

bool save_file(const std::vector<patch_byte> &pvec, std::string fname)
{
    FILE *f = fopen(fname.c_str(), "wb");
    if (!f)
    {
        cerr << "Cannot open file: " << fname << endl;
        return false;
    }

    size_t cnt = fwrite(pvec.data(), 1, pvec.size(), f);
    fclose(f);

    return cnt == pvec.size();
}

std::string compute_hash(const std::vector<patch_byte> &pvec)
{
    md5wrapper md5;
    return md5.getHashFromBytes(pvec.data(), pvec.size());
}

int main (int argc, char *argv[])
{
    if (argc <= 3)
    {
        cerr << "Usage: binpatch check|apply|remove <exe> <patch>" << endl;
        return 2;
    }

    std::string cmd = argv[1];

    if (cmd != "check" && cmd != "apply" && cmd != "remove")
    {
        cerr << "Invalid command: " << cmd << endl;
        return 2;
    }

    std::string exe_file = argv[2];
    std::vector<patch_byte> bindata;
    if (!load_file(&bindata, exe_file))
        return 2;

    BinaryPatch patch;
    if (!patch.loadDIF(argv[3]))
        return 2;

    BinaryPatch::State state = patch.checkState(bindata.data(), bindata.size());
    if (state == BinaryPatch::Conflict)
        return 1;

    if (cmd == "check")
    {
        switch (state)
        {
        case BinaryPatch::Unapplied:
            cout << "Currently not applied." << endl;
            break;
        case BinaryPatch::Applied:
            cout << "Currently applied." << endl;
            break;
        case BinaryPatch::Partial:
            cout << "Currently partially applied." << endl;
            break;
        default:
            break;
        }

        return 0;
    }
    else if (cmd == "apply")
    {
        if (state == BinaryPatch::Applied)
        {
            cout << "Already applied." << endl;
            return 0;
        }

        patch.apply(bindata.data(), bindata.size(), true);
    }
    else if (cmd == "remove")
    {
        if (state == BinaryPatch::Unapplied)
        {
            cout << "Already removed." << endl;
            return 0;
        }

        patch.apply(bindata.data(), bindata.size(), false);
    }

    if (!save_file(bindata, exe_file + ".bak"))
    {
        cerr << "Could not create backup." << endl;
        return 1;
    }

    if (!save_file(bindata, exe_file))
        return 1;

    cout << "Patched " << patch.entries.size()
         << " bytes, new hash: " << compute_hash(bindata) << endl;
    return 0;
}

File: library/lua/binpatch.lua
-- Simple binary patch with IDA dif file support.

local _ENV = mkmodule('binpatch')

local function load_patch(name)
    local filename = name
    if not string.match(filename, '[./\\]') then
        filename = dfhack.getHackPath()..'/patches/'..dfhack.getDFVersion()..'/'..name..'.dif'
    end

    local file, err = io.open(filename, 'r')
    if not file then
        if string.match(err, ': No such file or directory') then
            return nil, 'patch not found'
        end
    end

    local old_bytes = {}
    local new_bytes = {}

    for line in file:lines() do
        if string.match(line, '^%x+:') then
            local offset, oldv, newv = string.match(line, '^(%x+):%s*(%x+)%s+(%x+)%s*$')
            if not offset then
                file:close()
                return nil, 'could not parse: '..line
            end

            offset, oldv, newv = tonumber(offset,16), tonumber(oldv,16), tonumber(newv,16)
            if oldv > 255 or newv > 255 then
                file:close()
                return nil, 'invalid byte values: '..line
            end

            old_bytes[offset] = oldv
            new_bytes[offset] = newv
        end
    end

    return { name = name, old_bytes = old_bytes, new_bytes = new_bytes }
end

local function rebase_table(input)
    local output = {}
    local base = dfhack.internal.getImageBase()
    for k,v in pairs(input) do
        local offset = dfhack.internal.adjustOffset(k)
        if not offset then
            return nil, string.format('invalid offset: %x', k)
        end
        output[base + offset] = v
    end
    return output
end

local function rebase_patch(patch)
    local nold, err = rebase_table(patch.old_bytes)
    if not nold then return nil, err end
    local nnew, err = rebase_table(patch.new_bytes)
    if not nnew then return nil, err end
    return { name = patch.name, old_bytes = nold, new_bytes = nnew }
end

BinaryPatch = defclass(BinaryPatch)

BinaryPatch.ATTRS {
    name = DEFAULT_NIL,
    old_bytes = DEFAULT_NIL,
    new_bytes = DEFAULT_NIL,
}

function load_dif_file(name)
    local patch, err = load_patch(name)
    if not patch then return nil, err end

    local rpatch, err = rebase_patch(patch)
    if not rpatch then return nil, err end

    return BinaryPatch(rpatch)
end

function BinaryPatch:status()
    local old_ok, err, addr = dfhack.internal.patchBytes({}, self.old_bytes)
    if old_ok then
        return 'removed'
    elseif dfhack.internal.patchBytes({}, self.new_bytes) then
        return 'applied'
    else
        return 'conflict', addr
    end
end

function BinaryPatch:isApplied()
    return dfhack.internal.patchBytes({}, self.new_bytes)
end

function BinaryPatch:apply()
    local ok, err, addr = dfhack.internal.patchBytes(self.new_bytes, self.old_bytes)
    if ok then
        return true, 'applied the patch'
    elseif dfhack.internal.patchBytes({}, self.new_bytes) then
        return true, 'patch is already applied'
    else
        return false, string.format('conflict at address %x', addr)
    end
end

function BinaryPatch:isRemoved()
    return dfhack.internal.patchBytes({}, self.old_bytes)
end

function BinaryPatch:remove()
    local ok, err, addr = dfhack.internal.patchBytes(self.old_bytes, self.new_bytes)
    if ok then
        return true, 'removed the patch'
    elseif dfhack.internal.patchBytes({}, self.old_bytes) then
        return true, 'patch is already removed'
    else
        return false, string.format('conflict at address %x', addr)
    end
end

return _ENV
